src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResultBuilder.cs

```diff
@@ -0,0 +1,126 @@
+        /// </summary>
+        /// <param name="result">The instance to consume the errors from.</param>
+        public void AddResult(ValidationResult? result)
+        {
+            if (result?.ErrorMessage is not null)
+            {
+                Errors.Add(result.MemberNames.Any() ?
+                           $"{string.Join(MemberSeparatorString, result.MemberNames)}: {result.ErrorMessage}" :
+                           result.ErrorMessage);
+            }
+        }
+
+        /// <summary>
+        /// Adds any validation error carried by the enumeration of <see cref="ValidationResult"/> instances to this instance.
+        /// </summary>
+        /// <param name="results">The enumeration to consume the errors from.</param>
+        public void AddResults(IEnumerable<ValidationResult?>? results)
+        {
+            if (results != null)
+            {
+                foreach (var result in results)
+                {
+                    AddResult(result);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds any validation errors carried by the <see cref="ValidateOptionsResult"/> instance to this instance.
+        /// </summary>
+        /// <param name="result">The instance to consume the errors from.</param>
+        public void AddResults(ValidateOptionsResult result)
+        {
+            ThrowHelper.ThrowIfNull(result);
+
+            if (result.Failed)
+            {
+                if (result.Failures is null)
+                {
+                    Errors.Add(result.FailureMessage);
+                }
+                else
+                {
+                    // We are adding each failure separately to have the right failures count in _errors list.
+                    // Otherwise we could add result.FailureMessage as one failure containing all result failures.
+                    foreach (var failure in result.Failures)
+                    {
+                        if (failure is not null)
+                        {
+                            Errors.Add(failure);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Builds <see cref="ValidateOptionsResult"/> based on provided data.
+        /// </summary>
+        /// <returns>New instance of <see cref="ValidateOptionsResult"/>.</returns>
+        public ValidateOptionsResult Build()
+        {
+            if (_errors?.Count > 0)
+            {
+                return ValidateOptionsResult.Fail(_errors);
+            }
+
+            return ValidateOptionsResult.Success;
+        }
+
+        /// <summary>
+        /// Reset the builder to the empty state
+        /// </summary>
+        public void Clear() => _errors?.Clear();
+
+        private List<string> Errors
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Why AggressiveInlining?  Isn't this only used on failure paths?

Reply:
Does it hurt if we keep it? It is a small property which does not expect to bloat the code anyway.

Reply:
> Does it hurt if we keep it?

It forces the caller to be larger unnecessarily.  If the JIT wants to inline it, it can.  But this tells the JIT we really want it to be inlined, and in reality we actually would prefer it not be inlined as it's only for a failure code path.